[11.0][ADD] website_adv_image_optimization

tarteo opened the pull request:
This module allows the user to manually apply compression and resize options on an image.

simahawk commented on website_adv_image_optimization/controllers/main.py:
```diff
@@ -0,0 +1,55 @@
+        attachment_obj = request.env['ir.attachment']
+        attachment = attachment_obj.search([('id', '=', attachment_id)])
+
+        image = Image.open(io.BytesIO(base64.b64decode(attachment.datas)))
```
can't you use some tools from here https://github.com/odoo/odoo/blob/11.0/odoo/tools/image.py ?

tarteo replied:
Not really, because those do 1 operation at the time it will be less efficient. This way i can reuse some variables.

simahawk replied:
ok, then 3 recommendations: 
1. add a comment stating it (for me, for your future self and the ones that will ask the same question ;))
2. move resizing to a separated method 
3. close IO objects after getting the value

simahawk commented on website_adv_image_optimization/controllers/main.py:
```diff
@@ -0,0 +1,55 @@
+        attachment_obj = request.env['ir.attachment']
+        attachment = attachment_obj.search([('id', '=', attachment_id)])
+
+        image = Image.open(io.BytesIO(base64.b64decode(attachment.datas)))
```
As far as I see you use this only to get the image size to be used in `set_defaults`: can't you simply get the size via JS?

In any case you are leaving opened file objects around ;)

tarteo replied:
I can try

tarteo commented:
The travis error is related to https://github.com/OCA/website/issues/415

tarteo commented:
@simahawk Is this what you meant?

- I added a comment
- Moved resizing and changing jpeg quality to separated methods
- Closed Image object after all values are extracted

simahawk commented on website_adv_image_optimization/static/src/js/website_adv_image_optimization.js:
```diff
@@ -0,0 +1,107 @@
+            return done.promise();
+        },
+        apply: function() {
+        console.log(this);
```
debug bit lost? :)

simahawk commented on website_adv_image_optimization/tests/test_controller.py:
```diff
@@ -0,0 +1,32 @@
+
+
+class TestController(TransactionCase):
+    def test_optimize(self):
```
In the controller file you should import request:
```python
from odoo.http import request
```
then:
```python
import mock
[...]
    @mock.patch('dotted.path.to.your.controller.request')
    def test_optimize(self, mocked_request):
        mocked_request.env = self.env
        attachment = self.env.ref('website.business_conference')
        size = attachment.file_size
        ctrl = MainController()
        ctrl.optimize(attachment.id, 70, 300, 300)
        attachment = self.env.ref('website.business_conference')
        new_size = attachment.file_size
        self.assertFalse(new_size == size)
```

tarteo commented:
@simahawk Ah yes! Thanks!

Fuchur777 commented:
@simahawk changes done to your liking? Can it pass your review standards?

simahawk commented:
let's remember to squash commits before merge :wink:

tarteo commented:
@simahawk I squashed the commits :)
